File: Data Structures & Algorithms/first-missing-positive/submission-4.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        for i in range(len(nums)):
            if nums[i] <= 0 or nums[i] > len(nums):
                nums[i] = len(nums) + 1

        for i in range(len(nums)):
            value = abs(nums[i])

            if value > len(nums):
                continue

            value -= 1

            if nums[value] > 0:
                nums[value] = -nums[value]

        for i in range(len(nums)):
            if nums[i] > 0:
                return i + 1

        return len(nums) + 1

File: Data Structures & Algorithms/rotate-array/submission-4.java
public class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k %= n;
        int count = 0;

        for (int i = 0; count < n; i++) {
            int cur = i;
            int prev = nums[i];
            do {
                int nextIdx = (cur + k) % n;
                int tp = nums[nextIdx];
                nums[nextIdx] = prev;
                prev = tp;
                cur = nextIdx;
                count++;
            } while (i != cur);
        }
    }
}